### DISCUSSION TITLE
What could be the reason of OutOfMemory with that custom XCom back-end?

### DISCUSSION THREAD

**Discussion by makrushin-evgenii** [apache/airflow#16774] [Q&A]
I have XCom with following  implementation of methods:
```python
@staticmethod
def serialize_value(value: Any):
    drive2_hook = Drive2Hook()
    drive2_client = drive2_hook.get_client()

    drive2_document_name = f'{Drive2XCom.__ROOT_DOCUMENT}/{uuid.uuid4()}'
    drive2_document_content = pickle.dumps(value)

    print(f'Uploading of {len(drive2_document_content)} bytes to Drive2 document `{drive2_document_name}` started')
    Drive2XCom.__partial_upload(drive2_client, drive2_document_name, drive2_document_content)
    print(f'Uploading of {len(drive2_document_content)} bytes to Drive2 document `{drive2_document_name}` succeed')

    return BaseXCom.serialize_value(drive2_document_name)

@staticmethod
def deserialize_value(result):
    drive2_hook = Drive2Hook()
    drive2_client = drive2_hook.get_client()

    drive2_document_name = BaseXCom.deserialize_value(result)

    print(f'Downloading of Drive2 document `{drive2_document_name}` started')
    drive2_document_content = Drive2XCom.__download_content(drive2_client, drive2_document_name)
    print(f'Downloading of Drive2 document `{drive2_document_name}` succeed, '
          f'{len(drive2_document_content)} bytes received')

    return pickle.loads(drive2_document_content)
```

It is supposed to transfer data through external file storage "Drive2", and links to data - through a standard XCOM. But when loading a large array with the `deserialize_value` method, I get an exception:
```
[2021-07-02 12:32:04,664] {logging_mixin.py:104} INFO - Downloading of Drive2 document `/km/airflow/xcom/38cc3715-9020-4839-b6ba-737761cb882a` started
[2021-07-02 12:32:04,665] {transport.py:57} INFO - Sending POST https://api.testkontur.ru/auth/v5.13/authenticate-by-pass?login=metrika_airflow (body: 33)
[2021-07-02 12:32:04,694] {logging_mixin.py:104} WARNING - /home/airflow/.local/lib/python3.6/site-packages/urllib3/connectionpool.py:988 InsecureRequestWarning: Unverified HTTPS request is being made to host 'api.testkontur.ru'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
[2021-07-02 12:32:04,733] {logging_mixin.py:104} WARNING - /home/airflow/.local/lib/python3.6/site-packages/urllib3/connectionpool.py:988 InsecureRequestWarning: Unverified HTTPS request is being made to host 'api.dev.kontur'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
[2021-07-02 12:32:06,958] {logging_mixin.py:104} INFO - Downloading of Drive2 document `/km/airflow/xcom/38cc3715-9020-4839-b6ba-737761cb882a` succeed, 200200006 bytes received
[2021-07-02 12:32:57,006] {taskinstance.py:1455} ERROR - (psycopg2.errors.OutOfMemory) out of memory
DETAIL:  Failed on request of size 536870912 in memory context "ErrorContext".

[SQL: UPDATE rendered_task_instance_fields SET rendered_fields=%(rendered_fields)s WHERE rendered_task_instance_fields.dag_id = %(rendered_task_instance_fields_dag_id)s AND rendered_task_instance_fields.task_id = %(rendered_task_instance_fields_task_id)s AND rendered_task_instance_fields.execution_date = %(rendered_task_instance_fields_execution_date)s]
[parameters: {'rendered_fields': '{"op_args": [[121, 110, 99, 34, 70, 94, 138, 132, 215, 48, 252, 79, 111, 148, 179, 138, 26, 11, 49, 130, 79, 129, 210, 50, 166, 31, 219, 70, 63, 202, ... (457025494 characters truncated) ... 0, 2, 118, 244, 158, 147, 36, 157, 96, 227, 88, 98, 142, 93, 86, 140, 251, 9, 18, 30, 6, 56, 120, 224, 213, 88, 238, 56, 224, 244]], "op_kwargs": {}}', 'rendered_task_instance_fields_dag_id': 'test_dag', 'rendered_task_instance_fields_task_id': 'show_array', 'rendered_task_instance_fields_execution_date': DateTime(2021, 7, 1, 8, 30, 18, 731864, tzinfo=Timezone('UTC'))}]
(Background on this error at: http://sqlalche.me/e/13/e3q8)
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1278, in _execute_context
    cursor, statement, parameters, context
  File "/home/airflow/.local/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.OutOfMemory: out of memory
DETAIL:  Failed on request of size 536870912 in memory context "ErrorContext".
```

It seems that after loading the data, Airflow is trying to somehow use it to update the records in the metabase, which is why it crashes with an out-of-memory error. How can I avoid this?

**Comment by potiuk** [comment]
It is too big indeed. I do not think you should use XCom to pass a lot of data even if you use custom backend. XCom (in any backend) is supposed to only exchange meta-data not the actual data. XCom is simply not a good idea to pass data between tasks. 

Instead you should save the data to external storage (GCS/S3 for example) and only pass an id/url to that data via XCom.

What happened in this case -  It looks like you are trying to pass (and use) 0.5 GB of data from xcom and render it via Jinja Template. Your task is likely using xcom_pull to pull the data. This is likely op_args = "{{ ...xcom_pull() ...}}"

In this case Airlfow uses `rendered_task_instance` table to store the rendered value of the field (I guess the "op_args" is the guilty one) so that when you display it in the UI, you could see the actual value passed. And it's too much for the sqlalchemy to handle.

The solution - Just DON't use XCom to send a lot of data.

**Reply by makrushin-evgenii** [reply]
> Your task is likely using xcom_pull to pull the data. This is likely op_args = "{{ ...xcom_pull() ...}}"

In this case I use TaskFlow API. DAG code: 
```
with DAG(
        dag_id='test_dag',
        start_date=datetime.datetime(2021, 6, 1),
        catchup=False,
) as dag:
    @task
    def generate_array():
        from random import getrandbits
        return [getrandbits(8) for _ in range(100_000_000)]


    @task
    def save_array(array):
        drive2_hook = Drive2Hook()
        drive2_client = drive2_hook.get_client()

        drive2_client.upload_document('/km/airflow/xcom/test', str(array).encode(),
                                      replace_if_exists=True,
                                      days_to_live=1)

    @task
    def show_array(array):
        print(f'ARRAY LENGTH: {len(array)}, ARRAY FIRST ITEMS: {array[:10]}')


    array = generate_array()
    show_array(array)
    save_array(array)
```

> Instead you should save the data to external storage (GCS/S3 for example) and only pass an id/url to that data via XCom.

I would like not to repeat these steps every time in every method. It is not clear the difference between download explicitly in the task method itself and doing the same downloading implicitly with custom XCom

> In this case Airlfow uses rendered_task_instance table to store the rendered value of the field (I guess the "op_args" is the guilty one) so that when you display it in the UI, you could see the actual value passed. And it's too much for the sqlalchemy to handle.

Is it possible to disable this functionality?

**Reply by potiuk** [reply]
I took a look at the docs and I think you can use `orm_deserialize_value` method in your custom xcom backend. It should be written in the way that it does not render/read the expensive data, just short representation of it in cases like that . See  https://airflow.apache.org/docs/apache-airflow/stable/concepts/xcoms.html#custom-backends which is used when the value is rendered for UI purpose. 

Also if you want to learn more about the whole subject of sharing bigger data via custom xcom backends -there is a talk about it next.week at the Airflow Summit (next week):  
https://airflowsummit.org/sessions/2021/customizing-xcom-to-enhance-data-sharing-between-tasks/

**Reply by makrushin-evgenii** [reply]
> I took a look at the docs and I think you can use orm_deserialize_value method in your custom xcom backend. 

Somehow missed this point in the documentation. Thanks for the tip

> Also if you want to learn more about the whole subject of sharing bigger data via custom xcom backends -there is a talk about it next.week at the Airflow Summit (next week):

It will be very useful, I will definitely take a look. Thanks again for your help
